cedar-policy-validator/src/typecheck.rs

```diff
@@ -1857,22 +1960,89 @@ impl<'a> Typechecker<'a> {
             panic!("`typecheck_mul` called with an expression kind other than `MulByConst`");
         };
 
-        let ans_arg = self.expect_type(
-            request_env,
-            prior_eff,
-            arg,
-            Type::primitive_long(),
-            type_errors,
-        );
+        let ans_arg = self.expect_type(request_env, prior_eff, arg, Type::long_top(), type_errors);
         ans_arg.then_typecheck(|arg_expr_ty, _| {
-            TypecheckAnswer::success({
-                ExprBuilder::with_data(Some(Type::primitive_long()))
-                    .with_same_source_info(mul_expr)
-                    .mul(arg_expr_ty, *constant)
-            })
+            let (result_type, overflow_detected) =
+                if let Some((cba1, lb1)) = Self::arith_overflow_check_prepare_arg(&arg_expr_ty) {
+                    let min_multiplied_opt = i64::checked_mul(*constant, lb1.min);
+                    let max_multiplied_opt = i64::checked_mul(*constant, lb1.max);
+                    let (new_min_opt, new_max_opt) = if *constant < 0 {
+                        (max_multiplied_opt, min_multiplied_opt)
+                    } else {
+                        (min_multiplied_opt, max_multiplied_opt)
+                    };
+                    Self::arith_overflow_check_finish(
+                        new_min_opt,
+                        new_max_opt,
+                        cba1,
+                        type_errors,
+                        mul_expr,
+                    )
+                } else {
+                    (Type::long_any(), false)
+                };
+            let result_expr_type = ExprBuilder::with_data(Some(result_type))
+                .with_same_source_info(mul_expr)
+                .mul(arg_expr_ty, *constant);
+            if overflow_detected {
+                TypecheckAnswer::fail(result_expr_type)
+            } else {
+                TypecheckAnswer::success(result_expr_type)
+            }
         })
     }
 
+    fn arith_overflow_check_prepare_arg(
+        expr_ty: &Expr<Option<Type>>,
+    ) -> Option<(bool, &LongBounds)> {
+        if let Some(Type::Primitive {
+            primitive_type:
+                Primitive::Long(LongBoundsInfo {
+                    can_be_any: cba,
+                    bounds_opt: Some(lb),
+                }),
+        }) = expr_ty.data()
+        {
+            Some((*cba, lb))
+        } else {
+            None
+        }
+    }
+
+    // NOTE: This function currently has a rather ad-hoc subset of logic that
+    // was convenient to factor out of the overflow checks as they are currently
+    // structured. The code structure will probably have to change as we polish
+    // the overflow analysis.
+    fn arith_overflow_check_finish(
+        new_min_opt: Option<i64>,
+        new_max_opt: Option<i64>,
+        any_arg_can_be_any: bool,
+        type_errors: &mut Vec<TypeError>,
+        arith_expr: &Expr,
+    ) -> (
```

Comment:
[nit] I prefer returning a struct with named fields to returning a tuple that requires comments to explain the fields.